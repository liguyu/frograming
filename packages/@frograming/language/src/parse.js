import pegParser from './parser.peg';

export default frogCode => {
  const ast = pegParser.parse(frogCode);
  return ast;
};
