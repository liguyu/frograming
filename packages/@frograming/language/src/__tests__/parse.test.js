import { readFile, readdirSync } from 'fs-extra';
import { join, basename } from 'path';
import parse from '../parse';

const readdirFullSync = (dir) => readdirSync(dir).map(name => join(dir, name));

describe('parser', () => {
  readdirFullSync(join(__dirname, './frogram-specs/frogs'))
    .filter(n => /.frog$/.test(n))
    .forEach(path => it(basename(path), async () => {
      expect(parse(await readFile(path, 'utf8'))).toMatchSnapshot();
    }));
});
