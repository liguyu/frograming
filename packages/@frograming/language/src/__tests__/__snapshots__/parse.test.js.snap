// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser /Users/ycmjason/Development/frograming/packages/@frograming/language/src/__tests__/frogram-specs/frogs/01.frog 1`] = `
Object {
  "body": Array [
    Object {
      "name": "moveDown",
      "type": "command",
    },
    Object {
      "name": "moveUp",
      "type": "command",
    },
    Object {
      "body": Array [
        Object {
          "name": "moveDown",
          "type": "command",
        },
      ],
      "n": 2,
      "type": "LoopStatement",
    },
    Object {
      "condition": "!isCarUp&&isCarLeft||isLogUp",
      "elseBody": Array [
        Object {
          "name": "moveUp",
          "type": "command",
        },
      ],
      "ifBody": Array [
        Object {
          "name": "moveDown",
          "type": "command",
        },
        Object {
          "body": Array [
            Object {
              "name": "moveLeft",
              "type": "command",
            },
            Object {
              "name": "moveUp",
              "type": "command",
            },
            Object {
              "condition": "isCarUp",
              "elseBody": Array [
                Object {
                  "name": "moveLeft",
                  "type": "command",
                },
              ],
              "ifBody": Array [
                Object {
                  "name": "moveUp",
                  "type": "command",
                },
              ],
              "type": "IfStatement",
            },
          ],
          "n": 2,
          "type": "LoopStatement",
        },
      ],
      "type": "IfStatement",
    },
  ],
  "type": "frogram",
}
`;

exports[`parser /Users/ycmjason/Development/frograming/packages/@frograming/language/src/__tests__/frogram-specs/frogs/02.frog 1`] = `
Object {
  "body": Array [
    Object {
      "name": "moveDown",
      "type": "command",
    },
    Object {
      "name": "moveUp",
      "type": "command",
    },
    Object {
      "body": Array [
        Object {
          "condition": "!isLogUp",
          "elseBody": Array [
            Object {
              "name": "moveLeft",
              "type": "command",
            },
          ],
          "ifBody": Array [
            Object {
              "name": "moveUp",
              "type": "command",
            },
          ],
          "type": "IfStatement",
        },
      ],
      "n": 3,
      "type": "LoopStatement",
    },
    Object {
      "condition": "isLogUp",
      "type": "WaitStatement",
    },
    Object {
      "body": Array [
        Object {
          "name": "moveUp",
          "type": "command",
        },
      ],
      "condition": "!isLogLeft",
      "type": "WhileStatement",
    },
  ],
  "type": "frogram",
}
`;

exports[`parser /Users/ycmjason/Development/frograming/packages/@frograming/language/src/__tests__/frogram-specs/frogs/ifStatement.frog 1`] = `
Object {
  "body": Array [
    Object {
      "name": "moveLeft",
      "type": "command",
    },
    Object {
      "condition": "isLogDown",
      "elseBody": Array [
        Object {
          "name": "moveUp",
          "type": "command",
        },
      ],
      "ifBody": Array [
        Object {
          "name": "moveRight",
          "type": "command",
        },
      ],
      "type": "IfStatement",
    },
    Object {
      "condition": "isCarUp",
      "elseBody": Array [],
      "ifBody": Array [
        Object {
          "name": "moveDown",
          "type": "command",
        },
      ],
      "type": "IfStatement",
    },
    Object {
      "name": "moveRight",
      "type": "command",
    },
  ],
  "type": "frogram",
}
`;

exports[`parser /Users/ycmjason/Development/frograming/packages/@frograming/language/src/__tests__/frogram-specs/frogs/waitStatement.frog 1`] = `
Object {
  "body": Array [
    Object {
      "name": "moveUp",
      "type": "command",
    },
    Object {
      "condition": "isCarDown",
      "type": "WaitStatement",
    },
    Object {
      "condition": "isWallLeft&&isCarUp",
      "type": "WaitStatement",
    },
    Object {
      "name": "moveDown",
      "type": "command",
    },
  ],
  "type": "frogram",
}
`;

exports[`parser /Users/ycmjason/Development/frograming/packages/@frograming/language/src/__tests__/frogram-specs/frogs/whileStatement.frog 1`] = `
Object {
  "body": Array [
    Object {
      "name": "moveLeft",
      "type": "command",
    },
    Object {
      "body": Array [
        Object {
          "name": "moveUp",
          "type": "command",
        },
        Object {
          "name": "moveRight",
          "type": "command",
        },
        Object {
          "body": Array [
            Object {
              "name": "moveLeft",
              "type": "command",
            },
          ],
          "condition": "isLogUp||isLogLeft",
          "type": "WhileStatement",
        },
        Object {
          "name": "moveDown",
          "type": "command",
        },
      ],
      "condition": "isCarDown",
      "type": "WhileStatement",
    },
    Object {
      "body": Array [],
      "condition": "isWallLeft",
      "type": "WhileStatement",
    },
  ],
  "type": "frogram",
}
`;
